autowarefoundation/autoware.universe
# feat(operation_mode_transition_manager): add package to manage vehicle autonomous mode change

## Description

 - Add OperationModeTransitionManager that
    - checks the availability of the Autonomous mode change
    - publish the availability status
    - manage the operation_mode transition (Now only Manual <-> Autonomous is only supported)
    - send control_mode_request when the autonomous mode change is requested and the mode change is permitted.

## Related links



## Tests performed

Make sure related PRs are already merged in your testing environment:

- https://github.com/autowarefoundation/autoware.universe/pull/1061
- https://github.com/autowarefoundation/autoware.universe/pull/1198
- https://github.com/tier4/tier4_autoware_msgs/pull/36
- https://github.com/tier4/tier4_autoware_msgs/pull/41
- https://github.com/autowarefoundation/autoware.universe/pull/1244
- https://github.com/autowarefoundation/autoware.universe/pull/1201
- https://github.com/tier4/tier4_ad_api_adaptor/pull/51

1. run planning simulator, set initial pose and goal.
2. send control mode MANUAL to simulator vehicle
```
# command 0: MANUAL, 1: AUTO
ros2 service call /control/control_mode_request tier4_vehicle_msgs/srv/ControlModeRequest "{mode: {data: 0}}"
```
3. check if the control mode is now MANUAL
```
# report: 1: AUTO, 2: MANUAL
ros2 topic echo /vehicle/status/control_mode 
```
4. run controller to publish manual driving command and check the vehicle is moving following the trajectory
```
ros2 launch trajectory_follower_nodes simple_trajectory_follower.launch.xml external_target_vel:=-1.0
```
5. check if the AUTO mode is available (if the `/autoware/engage` is false and the vehicle is moving, make sure the AUTO is not available)
```
ros2 topic echo /control/is_autonomous_available
```
6. set `autoware/engage` = true (you can use api or send command directly)
```
ros2 topic pub /autoware/engage autoware_auto_vehicle_msgs/msg/Engage "{engage: 1}"
```
7. make sure the `is_autonomous_available` topic is true. (You can check the detailed info for the prohibit reason in debug topic).
```
ros2 topic echo /control/operation_mode_transition_manager/debug_info
```
8. send AUTO mode to vehicle and make sure the mode is changed to AUTO. The vehicle velocity will be changed to follow the reference trajectory speed.
```
# 1: DRIVER, 2: AUTONOMOUS
ros2 service call /api/autoware/set/operator tier4_external_api_msgs/srv/SetOperator "{mode: {mode: 2}}"
```

## limitation and future work


- write readme (will be in separate PR created soon)
- Implementation
  - Currently implemented as a node, but could be integrated into the `vehicle_cmd_gate` as a class in the future PR because of the strong connection to the `vehicle_cmd_gate`
  - add a check function if the filter in the `vehicle_cmd_gate` is not active when releasing a transition mode 
  - gate filter is not working when the mode is transitted from MANUAL to TRANSITION.

## Notes for reviewers



## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.


## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.


## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.


After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/
